firedrake/output.py

```diff
@@ -651,3 +754,304 @@ def write(self, *functions, **kwargs):
+    plex.create(comm=viewer.comm)
+    plex.setPlexName(name)
+    format = _get_format(viewer)
+    viewer.pushFormat(format=format)
+    if format == ViewerHDF5.Format.HDF5_PETSC:
+        sfXB = plex.topologyLoad(viewer)
+        plex.coordinatesLoad(viewer)
+        plex.labelsLoad(viewer)
+        plex.setOptionsPrefix("loaded_")
+        plex.viewFromOptions("-dm_view")
+    elif format == ViewerHDF5.Format.HDF5_XDMF:
+        plex.load(viewer=viewer)  # TODO: load labels, too.
+        plex.setOptionsPrefix("loaded_")
+        plex.viewFromOptions("-dm_view")
+        plex.interpolate()
+        plex.setOptionsPrefix("interpolated_")
+        plex.viewFromOptions("-dm_view")
+        sfXB = None
+    viewer.popFormat()
+    return plex, sfXB
+
+
+from firedrake.functionspacedata import get_global_numbering, get_shared_data_key
+def get_sectiondm_name(nodes_per_entity, real_tensorproduct):
+    return "_".join(["sectiondm"] +
+                    [str(n) for n in nodes_per_entity] +
+                    [str(real_tensorproduct)])
+
+
+
+def find_sectiondm_name(h5file, mesh_name, fs_name):
+    h5group = h5file[os.path.join("/topologies", mesh_name, "dms")]
+    for sectiondm_name in h5group:
+        if fs_name in h5group[os.path.join(sectiondm_name, PREFIX + "_function_spaces")]:
+            return sectiondm_name
+    raise ValueError(f"FunctionSpace {fs_name} not found.")
+
+
+def find_function_space_name(h5file, mesh_name, function_name):
+    # MixedFunctionSpace?
+    if PREFIX + "_mixed_function_spaces" in h5file["/topologies"]:
+        base_group = h5file[os.path.join("/topologies",
+                                         PREFIX + "_mixed_function_spaces")]
+        for mixed_fs_name in base_group:
+            if PREFIX + "_functions" in base_group[mixed_fs_name]:
+                if function_name in base_group[os.path.join(mixed_fs_name, PREFIX + "_functions")]:
+                    return mixed_fs_name
+    h5group = h5file[os.path.join("/topologies", mesh_name, "dms")]
+    for sectiondm_name in h5group:
+        if function_name in h5group[os.path.join(sectiondm_name, "vecs")]:
+            return h5group[os.path.join(sectiondm_name, "vecs")][function_name].attrs[PREFIX + "_function_space"]
+    raise ValueError(f"Function {function_name} not found.")
+
+
+class Dict(dict): pass
+
+
+class CheckpointFile(object):
+    _mesh_cache = weakref.WeakValueDictionary()
+
+    def __init__(self, filename, mode, comm=COMM_WORLD):
+        self.viewer = ViewerHDF5()
+        self.viewer.create(filename, mode=mode, comm=comm)
+        commkey = comm.py2f()
+        assert commkey != MPI.COMM_NULL.py2f()
+        self._meshes = self._mesh_cache.setdefault((filename, commkey), Dict())
+        self._function_spaces = {}
+        self._function_space_datas = {}
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+
+    def save_mesh(self, mesh, format=ViewerHDF5.Format.HDF5_PETSC):
```

Comment:
Will this support saving different topology information (i.e. different dmplexs) in the same hdf5 file?

Reply:
Yes, this will. `PETSc` currently writes all topology information in `/`, but, as soon as it is fixed, this branch will reflect it.